src/adapters/philips_hue/hub.rs

```diff
@@ -0,0 +1,111 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use serde_json;
+use std::sync::Arc;
+use std::thread;
+use std::time::Duration;
+use super::hub_api::HubApi;
+use super::{ HueAction, PhilipsHueAdapter, structs };
+use traits::Controller;
+use uuid::Uuid;
+
+pub struct Hub<C> {
+    pub adapter: PhilipsHueAdapter<C>,
+    pub id: String,
+    pub ip: String,
+    pub api: Arc<HubApi>,
+}
+
+impl<C: Controller> Hub<C> {
+    pub fn new(adapter: PhilipsHueAdapter<C>, id: &str, ip: &str) -> Self {
+        // Get API token from config store, default to a random UUID.
+        let token = adapter.controller.get_config().get_or_set_default(
+            "philips_hue",
+            &format!("token_{}", id),
+            &Uuid::new_v4().to_simple_string());
+        Hub {
+            adapter: adapter,
+            id: id.to_owned(),
+            ip: ip.to_owned(),
+            api: Arc::new(HubApi::new(id, ip, &token)),
+        }
+    }
+    pub fn start(&self) {
+        info!("Starting Hue Hub Service for {}", self.id);
+        let adapter = self.adapter.clone();
+        let id = self.id.clone();
+        let api = self.api.clone();
+
+        thread::spawn(move || {
```

Comment:
Right now, the main problem I see is that you cannot safely start an adapter (in particular for unit testing) as it will have side-effects until the process quits.

Regarding your questions.

> Are adapters expected to quit if no target devices are detected?

I don't think so.

> Should adapters leave a minimal detection stub running for when devices pop up?

Yes.

> What's the penalty on resources if many unused adapters keep lingering in the background?

Time will tell.

> Should the user instead send an event to a Foxbox that adapters to re-run discovery? What's that interface going to look like for adapters?

I believe that we should also support this. Maybe something as simple as

```rust
/// Try again to detect any new connected devices. This method is called by the user from the
/// UX.
fn discover(&self) {
  // By default, do nothing.
}
```